src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala

```diff
@@ -196,8 +196,14 @@ trait ContextErrors extends splain.SplainErrors {
         s"Implicit definition ${if (currentRun.isScala3) "must" else "should"} have explicit type${
           if (!inferred.isErroneous) s" (inferred $inferred)" else ""
         }"
-      if (currentRun.isScala3) cx.warning(tree.pos, msg, WarningCategory.Scala3Migration)
-      else cx.warning(tree.pos, msg, WarningCategory.OtherImplicitType)
+      // Assumption: tree.pos.focus points to the beginning of the name.
+      // DefTree doesn't give the name position; also it can be a synthetic accessor DefDef with only offset pos.
+      val namePos = tree.pos.focus.withEnd(tree.pos.point + tree.symbol.decodedName.length)
+      val action =
+        if (currentUnit.sourceAt(namePos) == tree.symbol.decodedName) runReporting.codeAction("insert explicit type", namePos.focusEnd, s": $inferred", msg)
+        else Nil
+      if (currentRun.isScala3) cx.warning(tree.pos, msg, WarningCategory.Scala3Migration, action)
+      else cx.warning(tree.pos, msg, WarningCategory.OtherImplicitType, action)
```

Comment:
There is the recent `namePos` attachment; that was so that diagnostics would cover just the name. Oh in fact that's why I just looking at `val yield`, I didn't bother with `val x,y,z = 42`. I did not consider synthetics at all.

Reply:
I even remembered `namePos`. Indeed it's not there on synthetic trees like accessor `DefDef`s. I should probably have tried harder to fix that instead of working around it...